src/components/TradeAssetSearch/hooks/useGetCustomTokensQuery.tsx

```diff
@@ -0,0 +1,62 @@
+import { type ChainId } from '@shapeshiftoss/caip'
+import { useQueries, type UseQueryResult } from '@tanstack/react-query'
+import type { TokenMetadataResponse } from 'alchemy-sdk'
+import { useCallback } from 'react'
+import { mergeQueryOutputs } from 'react-queries/helpers'
+import { isAddress } from 'viem'
+import { useFeatureFlag } from 'hooks/useFeatureFlag/useFeatureFlag'
+import { getAlchemyInstanceByChainId } from 'lib/alchemySdkInstance'
+
+type TokenMetadata = TokenMetadataResponse & {
+  chainId: ChainId
+  contractAddress: string
+  price: string
+}
+
+type UseGetCustomTokensQueryProps = {
+  contractAddress: string
+  chainIds: ChainId[]
+}
+
+type CustomTokenQueryKey = ['customTokens', string, ChainId]
+
+const getCustomTokenQueryKey = (contractAddress: string, chainId: ChainId): CustomTokenQueryKey => [
+  'customTokens',
+  contractAddress,
+  chainId,
+]
+
+export const useGetCustomTokensQuery = ({
+  contractAddress,
+  chainIds,
+}: UseGetCustomTokensQueryProps): UseQueryResult<(TokenMetadata | null)[], Error[]> => {
+  const customTokenImportEnabled = useFeatureFlag('CustomTokenImport')
+
+  const getTokenMetadata = useCallback(
+    async (chainId: ChainId) => {
+      const alchemy = getAlchemyInstanceByChainId(chainId)
+      const tokenMetadataResponse = await alchemy.core.getTokenMetadata(contractAddress)
+      // TODO: get price from somewhere
+      return { ...tokenMetadataResponse, chainId, contractAddress, price: '0' }
+    },
+    [contractAddress],
+  )
+
+  const getQueryFn = useCallback(
+    (chainId: ChainId) => () =>
+      isAddress(contractAddress, { strict: false }) ? getTokenMetadata(chainId) : null,
```

Comment:
One thing noted is we wont be caching errored responses, i.e no results from a chain will always re-fetch. We might want to consider caching these to reduce API quota usage

Reply:
Great minds ser: https://github.com/shapeshift/web/pull/7285#discussion_r1668177884 / https://github.com/shapeshift/web/pull/7285#discussion_r1668216261

Though to add nuance to the terminology used here:
- we *do* cache errors as errors with the current implementation (i.e if there is an upstream error, it will be handled the usual react-query way, though we should be cognicent that the default is 3 retries, so unless explicitly specifying a retry of 0, things may be quite long to be reflected in case of failures)
- this isn't an error per se but a no result case - so it does make sense not to run the query at all, however, there is an issue with the queryKey which is the reason why we're so overfetchy: